crates/aiken-lang/src/tipo/expr.rs

```diff
@@ -26,12 +29,126 @@ use crate::{
 use std::{cmp::Ordering, collections::HashMap, ops::Deref, rc::Rc};
 use vec1::Vec1;
 
+pub(crate) fn infer_function(
+    fun: &UntypedFunction,
+    module_name: &str,
+    hydrators: &mut HashMap<String, Hydrator>,
+    environment: &mut Environment<'_>,
+    lines: &LineNumbers,
+    tracing: Tracing,
+) -> Result<Function<Rc<Type>, TypedExpr, TypedArg>, Error> {
+    if let Some(typed_fun) = environment.inferred_functions.get(&fun.name) {
+        return Ok(typed_fun.clone());
+    };
+
+    let Function {
+        doc,
+        location,
+        name,
+        public,
+        arguments,
+        body,
+        return_annotation,
+        end_position,
+        can_error,
+        return_type: _,
+    } = fun;
+
+    let preregistered_fn = environment
+        .get_variable(name)
+        .expect("Could not find preregistered type for function");
+
+    let field_map = preregistered_fn.field_map().cloned();
+
+    let preregistered_type = preregistered_fn.tipo.clone();
+
+    let (args_types, return_type) = preregistered_type
+        .function_types()
+        .unwrap_or_else(|| panic!("Preregistered type for fn {name} was not a fn"));
+
+    // Infer the type using the preregistered args + return types as a starting point
+    let (tipo, arguments, body, safe_to_generalise) = environment.in_new_scope(|environment| {
+        let args = arguments
+            .iter()
+            .zip(&args_types)
+            .map(|(arg_name, tipo)| arg_name.to_owned().set_type(tipo.clone()))
+            .collect();
+
+        let hydrator = hydrators
+            .remove(name)
+            .unwrap_or_else(|| panic!("Could not find hydrator for fn {name}"));
+
+        let mut expr_typer = ExprTyper::new(environment, hydrators, lines, tracing);
+
+        expr_typer.hydrator = hydrator;
```

Comment:
Nothing is really changing in this function other than this bit. We now do pass the `hydrators` to the expression typer, to allow calling `infer_function` from within inferring an expression (e.g. a function body).